### ISSUE TITLE
Improve AST manipulation options

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#455]
I think there are some good use cases for working with strings from Ruby: 

- Programmatically building or modifying a query, as in graphql-client 
- Merging queries together as in #406 
- Handling subscriptions efficiently (eg, unioning all queries into one, resolving it once, then distrubuting the requested parts of the response to the various subscribers) 

However, this gem provides nothing for doing that. The best API would be one that doesn't reference AST node classes by name, instead, something decoupled from that, in case we have to change the nodes in the future. 

I heard good things about [estools](https://github.com/estools) a while back, there may be some inspiration there!

**Comment by rmosolgo** [comment]
GraphQL-JS supports this really nicely through their visitor API: http://graphql.org/graphql-js/language/#visit

For each kind of node, you may define an `enter()` and `leave()` function, and the return value of that function may affect the AST: 

```js
var editedAST = visit(ast, {
  enter(node, key, parent, path, ancestors) {
    // @return
    //   undefined: no action
    //   false: skip visiting this node
    //   visitor.BREAK: stop visiting altogether
    //   null: delete this node
    //   any value: replace this node with the returned value
  },
  leave(node, key, parent, path, ancestors) {
    // @return
    //   undefined: no action
    //   false: no action
    //   visitor.BREAK: stop visiting altogether
    //   null: delete this node
    //   any value: replace this node with the returned value
  }
});
```

I would love to do something like that, but to support immutable AST nodes, it could be a bit tricky!

**Comment by rmosolgo** [comment]
This will be available in 1.9 with #1740
